File: android/app/src/main/java/com/slackoff/app/playback/M3u8Parser.kt
package com.slackoff.app.playback

data class M3u8Segment(
    val id: String = java.util.UUID.randomUUID().toString(),
    val url: String,
    val duration: Double
)

object M3u8Parser {
    fun parse(text: String, baseUrl: String): List<M3u8Segment> {
        val segments = mutableListOf<M3u8Segment>()
        var pendingDuration = 0.0

        for (rawLine in text.lines()) {
            val line = rawLine.trim()
            if (line.isEmpty()) continue

            if (line.startsWith("#EXTINF:")) {
                val value = line.removePrefix("#EXTINF:").trim()
                pendingDuration = value.split(",").firstOrNull()?.toDoubleOrNull() ?: 0.0
            } else if (line.startsWith("#")) {
                continue
            } else {
                val url = resolve(line, baseUrl) ?: continue
                segments.add(M3u8Segment(url = url, duration = maxOf(pendingDuration, 1.0)))
                pendingDuration = 0.0
            }
        }
        return segments
    }

    fun resolve(value: String, baseUrl: String): String? {
        if (value.startsWith("/ets/")) {
            val decoded = decodeETSTail(value)
            if (decoded != null && decoded.startsWith("http")) return decoded
        }
        return if (value.startsWith("http://") || value.startsWith("https://")) {
            value
        } else {
            val base = baseUrl.substringBeforeLast("/")
            "$base/$value"
        }
    }

    private fun decodeETSTail(value: String): String? {
        val tail = value.split("/").lastOrNull() ?: return null
        var encoded = tail
        while (encoded.length % 4 != 0) encoded += "="
        return try {
            val decoded = android.util.Base64.decode(encoded, android.util.Base64.DEFAULT)
            String(decoded, Charsets.UTF_8)
        } catch (e: Exception) {
            null
        }
    }
}

File: SlackOff/Playback/M3U8Parser.swift
import Foundation

struct M3U8Segment: Identifiable, Equatable {
    let id = UUID()
    let url: URL
    let duration: Double
}

nonisolated enum M3U8Parser {
    static func parse(_ text: String, baseURL: URL) -> [M3U8Segment] {
        var segments: [M3U8Segment] = []
        var pendingDuration = 0.0

        // 注意：Swift 里 "\r\n" 是一个 Character，用 split(separator: "\n") 切不动 CRLF 播放列表
        for rawLine in text.split(whereSeparator: \.isNewline) {
            let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
            if line.isEmpty { continue }

            if line.hasPrefix("#EXTINF:") {
                let value = line.dropFirst("#EXTINF:".count).trimmingCharacters(in: .whitespaces)
                pendingDuration = Double(value.split(separator: ",").first ?? "") ?? 0
            } else if line.hasPrefix("#") {
                continue
            } else if let url = resolve(line, baseURL: baseURL) {
                segments.append(M3U8Segment(url: url, duration: max(pendingDuration, 1)))
                pendingDuration = 0
            }
        }
        return segments
    }

    static func resolve(_ value: String, baseURL: URL) -> URL? {
        if value.hasPrefix("/ets/") {
            if let decoded = decodeETSTail(value), let url = URL(string: decoded) {
                return url
            }
        }
        if let url = URL(string: value), url.scheme != nil {
            return url
        }
        return URL(string: value, relativeTo: baseURL)?.absoluteURL
    }

    private static func decodeETSTail(_ value: String) -> String? {
        let tail = value.split(separator: "/").last.map(String.init) ?? ""
        var encoded = tail
        while encoded.count % 4 != 0 { encoded += "=" }
        guard let data = Data(base64Encoded: encoded),
              let decoded = String(data: data, encoding: .utf8),
              decoded.hasPrefix("http") else { return nil }
        return decoded
    }
}
